pre_commit/languages/golang.py

```diff
@@ -17,20 +25,92 @@
 from pre_commit.util import rmtree
 
 ENVIRONMENT_DIR = 'golangenv'
-get_default_version = helpers.basic_get_default_version
 health_check = helpers.basic_health_check
 
 
-def get_env_patch(venv: str) -> PatchesT:
+@functools.lru_cache(maxsize=1)
+def get_default_version() -> str:
+    if helpers.exe_exists('go'):
+        return 'system'
+    else:
+        return C.DEFAULT
+
+
+def get_env_patch(venv: str, language_version: str) -> PatchesT:
+    if language_version == 'system':
+        return (
+            ('PATH', (os.path.join(venv, 'bin'), os.pathsep, Var('PATH'))),
+        )
+
+    return (
+        ('GOROOT', os.path.join(venv, '.go')),
+        (
+            'PATH', (
+                os.path.join(venv, 'bin'), os.pathsep,
+                os.path.join(venv, '.go', 'bin'), os.pathsep, Var('PATH'),
+            ),
+        ),
+    )
```

Comment:
All I can think of are some builtin commands that are useful to hookify (😆) and require access to a Go instance e.g. `gofmt`, `go generate`, `go mod ...` , `go vet`.
 
The question now is; do we want to give each an independent env? or expect/require the user to have a system installation? what if he uses a language version manager -- will it be cumbersome to active it?

If no to independent env, then we can yeet it.

Reply:
I've been thinking about this a bit -- and it probably benefits pre-commit to store languages separate from the environments maybe?  that way they can all be shared instead of having N copies of `go` -- I'm not sure what it would look like yet (probably another entry in the store database?)

Reply:
I had the same thought when I started working on this PR. It will mainly benefit the end user, thus pre-commit, especially when dealing with e.g. large sized languages in this case Go. Speaking of which, Go is a good candidate in this case, as it offers a simple way of achieving what we're looking for. I'm not sure if other languages offer the same experience -- I think that we can leverage `virtualenv` and `rbenv` for `python` and `ruby` -- if it's too complex or not worth it, we can have both structures coexist, maybe.

As a start, I think a very simple solution would be to pass a computed path to a language dir, install the language there, and make available a helper function that returns the language binary, this won't require a new database entry, but again this is a very simple solution based on my little knowledge of the tool internals.

Reply:
yeah I think I'll design something separate from this -- I'm imaginging any "first class" language will support this (so ruby, js, rust, and now go) -- and any second class language (python) will continue doing nothing.  I'll need to think of the right api for this though (and hopefully something that makes it ~easy for pre-commit.ci to use as well)

Reply:
Oh! python is a second class language, I forgot that, knowing that I've read it in the CONTRIBUTING docs